COMMAND:
    usage: EToKi.py clust [-h] -i INPUT -p PREFIX [-d IDENTITY] [-c COVERAGE]
                          [-t N_THREAD]

    Get clusters and exemplar sequences of the clusters. Run mmseqs linclust
    iteratively.

    optional arguments:
      -h, --help            show this help message and exit
      -i INPUT, --input INPUT
                            [INPUT; REQUIRED] name of the gene fasta file.
      -p PREFIX, --prefix PREFIX
                            [OUTPUT; REQUIRED] prefix of the outputs.
      -d IDENTITY, --identity IDENTITY
                            [PARAM; DEFAULT: 0.9] minimum intra-cluster identity.
      -c COVERAGE, --coverage COVERAGE
                            [PARAM; DEFAULT: 0.9] minimum intra-cluster coverage.
      -t N_THREAD, --n_thread N_THREAD
                            [PARAM; DEFAULT: 8] number of threads to use.


EXAMPLE:
    python EToKi.py clust -p examples/Escherichia.Achtman.alleles_clust -i examples/Escherichia.Achtman.alleles.fasta -d 0.95 -c 0.95

OUTPUT:
    2019-03-11 13:16:53.819085      Exemplar sequences in examples/Escherichia.Achtman.alleles_clust.clust.exemplar
    2019-03-11 13:16:53.819203      Clusters in examples/Escherichia.Achtman.alleles_clust.clust.tab

OUTPUT FILES:
head examples/Escherichia.Achtman.alleles_clust.clust.exemplar
>adk_1
GGGGAAAGGGACTCAGGCTCAGTTCATCATGGAGAAATATGGTATTCCGCAAATCTCCACTGGCGATATGCTGCGTGCTGCGGTCAAATCTGGCTCCGAGCTGGGTAAACAAGCAAAAGACATTATGGATGCTGGCAAACTGGTCACCGACGAACTGGTGATCGCGCTGGTTAAAGAGCGCATTGCTCAGGAAGACTGCCGTAATGGTTTCCTGTTGGACGGCTTCCCGCGTACCATTCCGCAGGCAGACGCGATGAAAGAAGCGGGCATCAATGTTGATTACGTTCTGGAATTCGACGTACCGGACGAACTGATTGTTGATCGTATCGTAGGCCGCCGCGTTCATGCGCCGTCTGGTCGTGTTTATCACGTTAAATTCAATCCGCCGAAAGTAGAAGGCAAAGACGACGTTACCGGTGAAGAACTGACTACCCGTAAAGACGATCAGGAAGAAACCGTACGTAAACGTCTGGTTGAATACCATCAGATGACTGCACCGCTGATCGGCTACTACTCCAAAGAAGCGGAAGCGGGTA
>adk_44
GGGTAAAGGAACTCAGGCTCAGTTCATCATGGAGAAATATGGTATTCCGCAAATCTCCACTGGCGATATGCTGCGCGCCGCAGTGAAATCAGGCTCCGAGTTGGGCAAACAGGCGAAAGATATCATGGACGCCGGTAAACTGGTGACCGATGAACTGGTGATTGCGCTGGTTAAAGAGCGTATCGCCCAGGAAGACTGCCGTAACGGTTTTCTGCTGGACGGTTTCCCGCGCACGATCCCGCAGGCTGACGCGATGAAAGAAGCGGGTATTGTGGTGGATTACGTGCTGGAATTCGACGTACCGGACGAACTGATCGTTGACCGTATTGTGGGTCGTCGCGTACACGCCGCCTCTGGCCGCGTTTACCACGTTAAGTTTAATCCGCCGAAAGTGGAAGGCAAAGATGACGTCACCGGCGAAGATCTGACGACCCGTAAAGACGATCAGGAAGAGACCGTTCGCAAACGTCTGGTGGAATATCATCAGATGACCGCGCCGCTGATTGGCTACTACCAGAAAGAAGCGGAAGCGGGCA
>adk_45
CGGGTAAGGGACTCAGGCTCAGTTCATCATGGAGAAGTACGGTATTCCGCAGATCTCCACCGGTGATATGCTTCGCGCCGCGGTAAAATCTGGCTCAGAGCTGGGTAAACAAGCAAAAGACATCATGGACGCCGGTAAACTGGTTACTGATGAACTGGTTATCGCGTTGGTGAAAGAACGCATTGCACAGGAAGATTGCCGCAACGGTTTCCTGCTGGACGGCTTCCCGCGTACTATTCCGCAGGCAGACGCCATGAAAGAAGCGGGTATCAACGTTGACTACGTACTGGAGTTCGACGTACCGGACGAGCTGATCGTTGACCGTATTATTGGTCGTCGCGTACACGCAGCGTCTGGCCGCGTTTACCACATCAAATTCAATCCGCCGAAGGTAGAAGGCAAAGACGACGTGACCGGCGAAGAGCTGACCACCCGTAAAGACGATCAGGAAGAGACCGTGCGTAAGCGCCTGGTGGAATACCATCAGATGACCGCGCCGCTGATTGGCTACTACTCCAAAGAAGCACAGGCGGGTA
>adk_89
GGGGAAAGGGACTCAGGCTCAGTTCATCATGGAGAAATATGGTATTCCGCAAATCTCCACAGGTGATATGCTGCGTGCCGCAGTCAAATCTGGCTCCGAGCTGGGTAAACAAGCAAAAGACATTATGGACGCTGGCAAACTGGTTACCGATGAACTGGTGATCGCGCTGGTTAAAGAACGTATTGCTCAGGAAGACTGCCGTAACGGTTTCCTGCTTGACGGCTTCCCGCGTACTATTCCGCAGGCTGACGCGATGAAAGAAGCGGGCATCAATGTTGATTACGTTCTGGAGTTCGACGTACCGGACGAACTGATTGTTGATCGTATTGTTGGTCGCCGCGTTCACGCGCCATCTGGCCGTGTTTATCACGTTAAATTCAATCCGCCGAAAGTAGAAGGCAAAGACGACGTTACTGGTGAAGACCTGACTACCCGTAAAGACGATCAGGAAGAAACTGTACGTAAACGTCTGGTCGAATACCATCAGATGACGGCGCCGCTGATCGGCTACTACTCTAAAGAAGCGGAAGCAGGCA
>adk_332
GGGTAAAGGAACTCAGGCTCAGTTCATCATGGAGAAGTATGGTATTCCGCAAATCTCCACCGGTGATATGCTGCGTGCCGCCGTGAAATCTGGCTCCGAGCTGGGCAAACAAGCGAAAGACATCATGGACGCCGGTAAGCTGGTGACTGATGAGCTGGTGATCGCGCTGGTAAAAGAGCGTATTGCTCAGGAAGACTGCCGTAATGGTTTCCTGCTTGATGGCTTCCCGCGTACTATTCCGCAGGCAGACGCCATGAAAGAAGCGGGCATCAACGTGGACTACGTTCTGGAGTTTGACGTACCTGACGAACTGATCGTTGACCGTATCGTCGGTCGCCGCGTACACGCTGCTTCTGGCCGCGTTTACCACATCAAATTCAACCCGCCTAAGGTTGAAGGCAAAGACGACGTGACCGGCGAAGAGCTGACCACCCGTAAAGACGATCAGGAAGAGACCGTACGTAAGCGCCTGGTGGAATACCATCAGATGACCGCTCCGCTGATCGGTTACTATTCTAAAGAAGCGCAGGCGGGTA

head examples/Escherichia.Achtman.alleles_clust.clust.tab
adk_1   adk_1
adk_10  adk_1
adk_100 adk_1
adk_101 adk_1
adk_103 adk_1
adk_108 adk_1
adk_109 adk_1
adk_11  adk_1
adk_111 adk_1
adk_112 adk_1
