COMMAND:
usage: EToKi.py MLSTdb [-h] -i ALLELEFASTA [-r REFSET] [-d DATABASE]
                       [-s REFSTRAIN] [-x MAX_IDEN] [-m MIN_IDEN] [-p PARALOG]
                       [-c COVERAGE] [-e]

MLSTdb. Create reference sets of alleles for nomenclature.

optional arguments:
  -h, --help            show this help message and exit
  -i ALLELEFASTA, --input ALLELEFASTA
                        [REQUIRED] A single file contains all known alleles in
                        a MLST scheme.
  -r REFSET, --refset REFSET
                        [DEFAULT: No ref allele] Output - Reference alleles
                        used for MLSType.
  -d DATABASE, --database DATABASE
                        [DEFAULT: No allele DB] Output - A lookup table of all
                        alleles.
  -s REFSTRAIN, --refstrain REFSTRAIN
                        [DEFAULT: None] A single file contains alleles from
                        the reference genome.
  -x MAX_IDEN, --max_iden MAX_IDEN
                        [DEFAULT: 0.9 ] Maximum identities between resulting
                        refAlleles.
  -m MIN_IDEN, --min_iden MIN_IDEN
                        [DEFAULT: 0.6 ] Minimum identities between refstrain
                        and resulting refAlleles.
  -p PARALOG, --paralog PARALOG
                        [DEFAULT: 0.1 ] Minimum differences between difference
                        loci.
  -c COVERAGE, --coverage COVERAGE
                        [DEFAULT: 0.7 ] Proportion of aligned regions between
                        alleles.
  -e, --relaxEnd        [DEFAULT: False ] Allow changed ends (for pubmlst).


EXAMPLE:
python EToKi.py MLSTdb -i examples/Escherichia.Achtman.alleles.fasta -r examples/Escherichia.Achtman.references.fasta -d examples/Escherichia.Achtman.convert.tab

OUTPUTS:
2019-03-11 15:51:06.654794      identifed 133 good exemplar alleles after nucleic search
2019-03-11 15:51:09.012531      identifed a total of 137 good exemplar alleles after amino search
2019-03-11 15:51:11.135598      There are 33 good exemplar alleles left after final clustering
2019-03-11 15:51:11.142601      A file of reference alleles has been generated:  examples/Escherichia.Achtman.references.fasta
2019-03-11 15:51:11.239984      A lookup table of all alleles has been generated:  examples/Escherichia.Achtman.convert.tab

OUTPUT FILES:
head examples/Escherichia.Achtman.references.fasta
>adk_1
GGGGAAAGGGACTCAGGCTCAGTTCATCATGGAGAAATATGGTATTCCGCAAATCTCCACTGGCGATATGCTGCGTGCTGCGGTCAAATCTGGCTCCGAGCTGGGTAAACAAGCAAAAGACATTATGGATGCTGGCAAACTGGTCACCGACGAACTGGTGATCGCGCTGGTTAAAGAGCGCATTGCTCAGGAAGACTGCCGTAATGGTTTCCTGTTGGACGGCTTCCCGCGTACCATTCCGCAGGCAGACGCGATGAAAGAAGCGGGCATCAATGTTGATTACGTTCTGGAATTCGACGTACCGGACGAACTGATTGTTGATCGTATCGTAGGCCGCCGCGTTCATGCGCCGTCTGGTCGTGTTTATCACGTTAAATTCAATCCGCCGAAAGTAGAAGGCAAAGACGACGTTACCGGTGAAGAACTGACTACCCGTAAAGACGATCAGGAAGAAACCGTACGTAAACGTCTGGTTGAATACCATCAGATGACTGCACCGCTGATCGGCTACTACTCCAAAGAAGCGGAAGCGGGTA
>adk_212
GGGTAAAGGAACTCAGGCTCAGTTCATCATGGAGAAATATGGTATTCCGCAAATCTCCACTGGCGATATGCTGCGCGCCGCAGTGAAATCAGGCTCCGAGTTGGGCAAACAGGCGAAAGATATCATGGACGCCGGTAAACTGGTGACCGATGAACTGGTGATTGCGCTGGTTAAAGAGCGTATCGCCCAGGAAGACTGCCGTAACGGTTTTCTGCTGGACGGTTTCCCGCGCACGATCCCGCAGGCTGACGCAATGAAAGAAGCGGGTATTGTCGTGGATTACGTGCTGGAATTCGACGTACCGGACGAACTGATCGTTGACCGTATTGTGGGTCGTCGCGTACACGCCGCCTCTGGCCGCGTTTACCACGTTAAGTTTAATCCGCCGAAAGTAGAAGGCAAAGATGACGTCACCGGTGAAGATCTGACGACCCGTAAAGACGATCAGGAAGAGACCGTTCGCAAACGTCTGGTGGAATATCATCAGATGACCGCGCCGCTGATTGGCTACTACCAGAAAGAAGCGGAAGCGGGCA
>adk_424
GGGTAAAGGAACTCAGGCTCAGTTCATCATGGAGAAATACGGTATTCCGCAAATCTCCACCGGTGACATGCTGCGCGCCGCTGTTAAATCTGGCTCCGAGCTGGGTAAACAAGCGAAAGACATCATGGACGCAGGCAAACTGGTGACCGACGAACTGGTTATCGCGCTGGTTAAAGAGCGCATTGCTCAGGAAGACTGCCGTAACGGTTTCCTGCTGGACGGCTTCCCGCGCACCATTCCTCAGGCTGACGCCATGAAAGAAGCGGGCATCAACGTGGATTACGTGCTGGAGTTCGACGTACCAGACGAGCTGATCGTTGACCGTATCGTTGGCCGTCGCGTACACGCTGCTTCTGGCCGCGTTTACCACATCAAATTCAACCCACCTAAGGTTGAAGGCAAAGACGACGTGACCGGCGAAGAGCTGACCACCCGTAAAGACGATCAGGAAGAAACCGTGCGTAAACGCCTGGTGGAATATCATCAGATGACCGCACCGCTGATCGGCTACTACACCAAAGAAGCGCAGGCGGGTA
>fumC_257
CGAGAGCCATTCGCCAGGCGGCGGATGAAGTGCTGGCAGGGCAACATGATGATGAGTTCCCGCTGGCAATCTGGCAAACGGGATCCGGTACGCAAAGCAATATGAATATGAATGAAGTACTGGCTAATCGGGCCAGCGAATTGCTTGGCGGCGGCGTACGCGGAATGGAACGTAAAGTTCACCCTAACGATGACGTGAACAAAAGCCAAAGCTCCAATGATGTCTTCCCAACTGCTATGCACGTAGCGGCGTTATTGGCACTGCGTAAGCAACTCATTCCTCAGCTCAAAAACCTCACTCAAACTCTGAATGAGAAATCGCGTGCATTTGCCGATATCGTCAAAATTGGTCGTACCCACTTGCAGGACGCCACACCGCTGACCCTGGGACAGGAGATTTCTGGTTGGGTAGCGATGCTGGAGCATAATCTCAAACATATTGAATACAGCCTGCCTCACGTAGCGGAACTGGC
>fumC_464
CCACCGCCATCATCAATGCCGCCGACGAGGTGCTGGCGGGCAAACATCCCGACGAATTCCCCCTTGCCATCTGGCAGACCGGCTCAGGCACCCAGAGCAACATGAACATGAACGAGGTGCTGGCGAACCGCGCAAGCGAGCTTCTTGGCGGCGTGCGCGGCATGGAGCGCAAGGTCCACCCGAACGATGACGTGAACAAAAGCCAGAGTTCCAACGACGTGTTCCCGACGGCCATGCACGTGGCGGCCGTCATCGCAATCCGCGAGCAGCTTATCCCGCAGCTCAACGTGCTTAAATCTACGCTTAACGAGAAAGCGCAGGCCTTCCGCGACATCGTTAAAATTGGTCGTACGCACCTTCAGGACGCCACGCCGCTGACGCTTGGTCAGGAGATCTCCGGCTGGGTGGCAATGCTGGAGCACAATCTTAAGCATATCGACAACAGCCTGCCGCATCTGGCGGAGCTGGC

head examples/Escherichia.Achtman.convert.tab
13e727e8-e700-095a-08b8-4958fb5f29fe,adk,1
5054446f-a95f-3bf5-8f3a-97e52757e5a0,adk,2
03896ceb-fd69-303d-e8e5-b967ce11d204,adk,3
ec802d3a-a3d4-3249-2dcd-29b3904ff262,adk,4
69c972d9-18d5-47d8-6fdd-cc2c16b9cf2b,adk,5
978b57ff-073a-236f-21e3-74c65d92885d,adk,6
6b514679-5f15-2072-dbfd-c0b2f1f542bd,adk,7
19f3188a-e53a-55b9-57d6-10a023a441ee,adk,8
b369b4b7-6214-1d49-3ce6-599c124e9d43,adk,9
ff1dd41e-23a0-fa72-e129-e5968a6e66d2,adk,10
